Lib/multiprocessing/process.py

```diff
@@ -61,7 +61,7 @@ def parent_process():
 def _cleanup():
     # check for processes which have finished
     for p in list(_children):
-        if p._popen.poll() is not None:
+        if p._popen is not None and p._popen.poll() is not None:
```

Comment:
Since you were encountering a concurrency error where p._open was None, this change does not actually fix it yet, it just narrows the window.  The value could change between the is not None check and the next reference.

Instead, use something like this, it does the attribute lookup on p _a single time_ so there is no race condition.

```python
if (child_popen := p._popen) and child_popen.poll() is not None:
    _children.discard(p)
```

This still assumes that keeping the `multiprocess.popen_*.Popen` instance alive and calling poll on it after the process has already been observed to exit is okay, but that is likely true (it should just take the fast path and return the saved returncode).  _children is a set, it is safe to discard() something from a set that doesn't exist.

Reply:
Agreed, never used the walrus operator before, now I do see how valuable it is.